### ISSUE TITLE
Test a POST with MultiPart using TestApplicationEngine does not success or fail

### ISSUE THREAD

**Issue by razvn** [ktorio/ktor#1936]
Ktor 1.3.2, TestApplicationEngine 

When running a test that sends  post with multipart it runs forever and never stops.

No problem when server really runs. It's just the test.

**To Reproduce**

Server:
```kotlin
fun main(args: Array<String>): Unit = io.ktor.server.netty.EngineMain.main(args)

@Suppress("unused") // Referenced in application.conf
@kotlin.jvm.JvmOverloads
fun Application.module(testing: Boolean = false) {
    install(ContentNegotiation) {
    }

    routing {
        post("/") {
            val mp = try {
                call.receiveMultipart()
            } catch (e: Exception) {
                call.application.log.error("Getting multipart error"")
                null
            }

           if (mp == null) call.respond(HttpStatusCode.BadRequest, "Null value")
           else call.respond(HttpStatusCode.OK,"OK")
        }
    }
}
```
Test:
```kotlin
internal class ApplicationKtTest {
    private val boundary = "***bbb***"
    private val  sysTempDir = System.getProperty("java.io.tmpdir") ?: "/tmp"

    private val multipart = listOf(PartData.FileItem({ byteArrayOf(1, 2, 3).inputStream().asInput() }, {}, headersOf(
            HttpHeaders.ContentDisposition,
            ContentDisposition.File
                    .withParameter(ContentDisposition.Parameters.Name, "file")
                    .withParameter(ContentDisposition.Parameters.FileName, "test.jpg")
                    .toString()
    )))

    @Test
    fun testUploadApplication() = testApp {
        handlePost("/", boundary, multipart).apply {
            println("============= RESPONSE ====================")
            println(response.content)
            println("=================================")
        }
    }
}

private fun TestApplicationEngine.handlePost(uri: String,
         boundary: String,
         multipart: List<PartData.FileItem>,
         setup: TestApplicationRequest.() -> Unit = {}
): TestApplicationCall {
    return handleRequest(method = HttpMethod.Post, uri = uri)  {
        addHeader(HttpHeaders.ContentType,
                ContentType.MultiPart.FormData.withParameter("boundary", boundary).toString()
        )
        setBody(boundary, multipart)
        setup()
    }
}

private fun testApp(callback: TestApplicationEngine.() -> Unit): Unit {
    withTestApplication({ module(true) }, callback)
}
```
The test runs forever and gives no result. if `addHeader` and `setBody` are commented there is a response.

**Expected behavior**
The test should print `OK`

**Repository**
[If you just want to clone](https://github.com/razvn/ktor_test_multipart_error_exemple)  a test repository.

**Comment by neelkamath** [comment]
I have the same issue with my HTTP PATCH request.
@razvn do you have a workaround? I didn't understand your second comment.

**Comment by razvn** [comment]
@neelkamath just doing 
```kotlin
call.receiveMultipart().forEachPart { it.dispose }
```
is enough to make it work.

**Comment by neelkamath** [comment]
There are currently two workarounds. Note that these snippets are pseudocode.

## First Workaround

@razvn's workaround is to use `call.receiveMultipart().forEachPart()`.

So, this code **won't** work:
```kotlin
call.receiveMultipart().readPart()!!.dispose()
call.respond(HttpStatusCode.NoContent)
```

But this code **will** work:
```kotlin
call.receiveMultipart().forEachPart { it.dispose() }
call.respond(HttpStatusCode.NoContent)
```

## Second Workaround

I was wondering if this was because the call waits for some sort of EOF to know when to stop executing, but couldn't find a `close` method on `call.receiveMultipart()`. However, I was able to get rid of the call to `call.receiveMultipart().forEachPart()`. For example:
```kotlin
call.receiveMultipart().readPart()!!.dispose()
multipart.readPart() // Read it an extra time even though there's only one part.
call.respond(HttpStatusCode.NoContent)
```

**Comment by oleg-larshin** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/KTOR-345">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.

**Comment by e5l** [comment]
Fixed in 1.5.2
